packages/controls/src/controls/rich-text/v2/rich-text.ts

```diff
@@ -0,0 +1,166 @@
+  static readonly Mode = {
+    Inline: `${this.type}::mode::inline`,
+    Block: `${this.type}::mode::block`,
+  } as const
+
+  static get schema() {
+    const type = z.literal(this.type)
+    const version = z.literal(2)
+
+    const dataV1 = DTOSchema.valueJSON
+    const dataV2 = z.object({
+      type,
+      version,
+      descendants: z.array<SchemaType<Slate.Descendant>>(z.any()),
+      key: z.string(),
+    })
+
+    const data = z.union([dataV1, dataV2])
+    const value = dataV2
+
+    const userConfig = z.object({
+      mode: z.optional(
+        z.union([z.literal(this.Mode.Inline), z.literal(this.Mode.Block)]),
+      ),
+      defaultValue: z.optional(z.string()),
+    })
+
+    return {
+      type,
+      data,
+      dataV1,
+      dataV2,
+      value,
+      userConfig,
+    }
+  }
+
+  get controlType() {
+    return Definition.type
+  }
+
+  get schema() {
+    const baseSchema = Definition.schema
+    const definition = z.object({
+      type: baseSchema.type,
+      config: this.configSchema,
+    })
+
+    return {
+      ...baseSchema,
+      definition,
+      resolvedValue: this.nodeSchema,
+    }
+  }
+
+  get configSchema(): SchemaType<Config> {
+    return z.any()
+  }
+
+  get nodeSchema(): SchemaType<RuntimeNode> {
+    return z.any()
+  }
+
+  safeParse(data: unknown | undefined): ParseResult<DataType | undefined> {
+    return safeParse(this.schema.data, data)
+  }
+
+  fromData(data: DataType | undefined): ValueType | undefined {
+    return data !== undefined ? Definition.normalizeData(data) : undefined
+  }
+
+  toData(value: ValueType): DataType {
+    return value
+  }
+
+  copyData(
+    data: DataType | undefined,
+    context: CopyContext,
+  ): DataType | undefined {
+    if (data == null) return data
+
+    const nodes = Definition.dataToNodes(data)
+    return Definition.nodesToDataV2(copyNodes(nodes, context))
+  }
+
+  introspect<R>(
+    data: DataType | undefined,
+    target: IntrospectionTarget<R>,
+  ): R[] {
+    if (data == null) return []
+
+    const nodes = Definition.dataToNodes(data)
+    return introspectNodes(nodes, target, this.pluginControls)
+  }
+
+  static isV1Data(data: DataType | undefined): data is DataV1Type {
+    return isRichTextDTO(data)
+  }
+
+  static dataToNodes(data: DataType): Slate.Descendant[] {
+    return Definition.isV1Data(data) ? richTextDTOtoDAO(data) : data.descendants
+  }
+
+  static nodesToDataV2(
+    descendants: Slate.Descendant[],
+    key?: string,
+  ): DataV2Type {
+    return {
+      type: Definition.type,
+      version: 2,
+      descendants,
+      key: key ?? uuid(),
+    }
+  }
+
+  static normalizeData(data: DataType): DataV2Type {
+    return Definition.isV1Data(data)
+      ? Definition.nodesToDataV2(richTextDTOtoDAO(data))
+      : data
+  }
+
+  abstract toText(data: DataType | undefined): string
+  abstract get pluginControls(): RichTextPluginControl[]
+  abstract pluginControlAt(index: number): RichTextPluginControl | undefined
```

Comment:
Actual `RichTextV2` plugins depend on React, but the introspection and data copying code only needs access to plugin controls, so we're providing access to the latter through these abstract methods.